### PR TITLE
[RESOLVE #946] Fixing bug preventing StackGroup dependencies

### LINKED ISSUE #946: stack dependencies in cascading configs
assuming i have the following project structure

```
.
├── config.yaml
├── prod
│   ├── config.yaml
│   ├── vpc.yaml
│   └── vpcPeering.yaml
├── test
│   ├── config.yaml
│   ├── vpc.yaml
│   └── vpcPeering.yaml
├── vpc.yaml
└── vpcPeering.yaml
```

i would like to define a dependency in vpcPeering.yaml to vpc.

at the moment the only way i could get this work is to define the dependencies in both ./test/vpcPeering.yaml and ./prod/vpcPeering.yaml

by adding 

```
dependencies:
    - {{ stack_group_config.profile }}/vpc.yam
```

is there a smarter way to do this. ideally i want to configure the dependency just 1 time in ./vpcPeering.yaml but i was not able to do so.

### PR DESCRIPTION
**Pull request by jfalkenstein**
This resolves issue #946 by solving the issue of inherited dependencies. Turns out, the problem was just a small bug with inherited objects in memory.

There was a small bug in the ConfigReader that has to do with the way list-joining works on individual stacks and how dependencies are resolved.

The basic situation was this:

* Sceptre uses the "list_join" strategy to combine StackGroup dependencies with Stack dependencies. 
* The list_join strategy creates a NEW list by combining A + B... **but only if both A and B exist**. 
* In the case where the StackGroup dependencies exist _but there are no Stack dependencies_, Sceptre just selects A and passes each stack in that group the exact same dependencies list (i.e. **same object in memory**) from the StackGroup
* The process of resolving stacks replaces the stack string with a Stack object _in place_.
* If two stacks have inherited the exact same dependencies list because neither of them defined their own dependencies lists, the first stack to be resolved will replace those objects in memory with Stack objects and the second stack to be resolved will receive those Stack objects rather than strings.
* In the attempt to resolve the Stack object (believing it's a string), Sceptre will blow up because you cannot call `.replace()` on a Stack instance. But we don't need to do that at all if it's already a Stack.

This PR fixes that bug by skipping resolution for dependencies that are already Stack objects.
## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [x] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,6 +11,7 @@ __pycache__/
 env/
 venv/
 .venv/
+.python-version
```
Comment (craighurley) [inline .gitignore:14]: Why was this included?

Comment (craighurley) [reply .gitignore:14]: I could have sworn I saw folks discussing adding adding that file to the repo, but it looks like it is not there, adding to ignore might be ok

Comment (jfalkenstein) [reply .gitignore:14]: This is used by pyenv to set up a virtual environment. It's similar to a venv directory. Nothing specific to the project, just prevents accidentally committing this to the local repo.

Comment (craighurley) [reply .gitignore:14]: I know the purpose of `.python-version`, my point is that it has been included in this repo in the past and I know folks discussed keeping it in the past.

https://github.com/Sceptre/sceptre/tree/20579c49ac62ec720cfedca097bbe55daf825a93
https://github.com/Sceptre/sceptre/search?q=%22.python-version%22&type=commits

Ideally a PR should only touch a limited set of files related to the functionality being added/removed/updated.

Comment (jfalkenstein) [reply .gitignore:14]: That's fine. I can pull it out. Not a big deal.

Comment (jfalkenstein) [reply .gitignore]: This has been pulled out.

**Comment by jfalkenstein** [comment]
@craighurley any chance I could get a follow-up review on this?

**Review by zaro0508 — CHANGES_REQUESTED** [review]
Issue #946 seems to indicate that this fix will provide new functionality to sceptre.  would you be willing to add some documentation telling users about cascading dependencies and how they configure sceptre for it?

File: tests/test_config_reader.py [hunk 2]
```diff
@@ -525,6 +525,39 @@ def test_missing_dependency(
...
+                "region": "region",
+                "template_path": rel_path,
+            }
+
+            abs_path = os.path.join(config_dir, rel_path)
+            self.write_config(abs_path, config)
+        self.context.project_path = project_path
+        try:
+            config_reader = ConfigReader(self.context)
+            all_stacks, command_stacks = config_reader.construct_stacks()
```
[the remark this answers is not in the document]
Comment (jfalkenstein) [reply tests/test_config_reader.py:555]: So we already test stack resolution in other tests (see here: https://github.com/Sceptre/sceptre/pull/1116/files#diff-b2719a0eb862c151c5232ec3d1d3f09086a76ad10a01f5dfa26deb37085b770eL465-L493). This particular test is covering the bug. With the code currently in master, this test fails. With my fix, it passes. I feel like that's sufficient.

**Comment by jfalkenstein** [comment]
@zaro0508 That's an interesting thought. #946 seems to call cascading dependencies as a feature, but in reality, there's nothing new about what I have done. In fact, we were already using the "list_join" strategy on dependencies. Sceptre cascades almost all configs and there was explicit code in place to handle this in the case of dependencies... but in specific circumstances, this cascade raised an error. So I would categorize this change more as a "bug fix" than a new feature.

With that said, I could absolutely submit a follow-up PR to improve the docs on "[Cascading Config](https://sceptre.cloudreach.com/2.6.3/docs/stack_group_config.html#cascading-config)", since Sceptre's cascading is one of its best features and it doesn't really ever highlight how awesome and flexible this feature-set is. I think adding docs that indicate the merge strategies for each stack configuration would also be super helpful, since you have to consult the source code to find that. Would you accept a follow-up PR that does this, @zaro0508 ?

File: tests/test_config_reader.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zaro0508) [reply tests/test_config_reader.py:555]: sounds good

**Review by zaro0508 — APPROVED** [review]
> Would you accept a follow-up PR that does this, @zaro0508 ?

Yes, that would be great @jfalkenstein!
